### ISSUE TITLE
[Clouddriver] Kubernetes - HPA can't get pod metrics for Kubernetes v1.8++

### ISSUE THREAD

**Issue by antoooks** [spinnaker/spinnaker#2249]
## Title ##
*[Clouddriver] Kubernetes - HPA can't get pod metrics for Kubernetes v1.8++*

## Cloud Provider ##
*Kubernetes*

## Environment ##
I am running Spinnaker v1.5.4 using Kubernetes v1.9 with AWS ECR. 

I am aware that Kubernetes 1.8++ doesn't use heapster anymore for metrics collection but metrics-server, so I got metrics-server pod running instead of heapster.

## Feature Area ##
Pipeline Stage "Deploy", Horizontal Pod Autoscaler (hpa) specs generated by the UI is malformed, causing hpa can't found its target pod.

## Description ##
I'm trying to enable autoscaling for our pods by using Spinnaker autoscaling feature, the hpa was created, also the pods. But, the hpa can't get the metrics out of the pods and shows <unknown> for CPU usage, thus the hpa doesn't work properly. It's not happening when I generate the hpa manually with kubectl command.

Here are the specs I got when using ```kubectl get hpa my-app -o yaml``` for both hpa generated by Spinnaker and kubectl command.

**With Spinnaker UI (actual spec) :**
```
apiVersion: v1
items:
- apiVersion: autoscaling/v1
  kind: HorizontalPodAutoscaler
  metadata:
    annotations:
      autoscaling.alpha.kubernetes.io/conditions: '[{"type":"AbleToScale","status":"False","lastTransitionTime":"2018-01-15T04:06:24Z","reason":"FailedGetScale","message":"the
        HPA controller was unable to get the target''s current scale: no matches for
        /, Kind=Deployment"}]'
    creationTimestamp: 2018-01-15T04:05:54Z
    name: my-app
    namespace: default
    resourceVersion: "12398720"
    selfLink: /apis/autoscaling/v1/namespaces/default/horizontalpodautoscalers/my-app
    uid: 620cb23c-f9a9-11e7-8056-0649c28415ba
  spec:
    maxReplicas: 3
    minReplicas: 1
    scaleTargetRef:
      kind: Deployment
      name: my-app
    targetCPUUtilizationPercentage: 50
  status:
    currentReplicas: 0
    desiredReplicas: 0
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**With kubectl command (expected spec):**
```
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  annotations:
    autoscaling.alpha.kubernetes.io/conditions: '[{"type":"AbleToScale","status":"False","lastTransitionTime":"2018-01-15T04:50:04Z","reason":"BackoffBoth","message":"the
      time since the previous scale is still within both the downscale and upscale
      forbidden windows"},{"type":"ScalingActive","status":"True","lastTransitionTime":"2018-01-15T04:49:34Z","reason":"ValidMetricFound","message":"the
      HPA was able to succesfully calculate a replica count from cpu resource utilization
      (percentage of request)"},{"type":"ScalingLimited","status":"False","lastTransitionTime":"2018-01-15T04:49:34Z","reason":"DesiredWithinRange","message":"the
      desired count is within the acceptable range"}]'
    autoscaling.alpha.kubernetes.io/current-metrics: '[{"type":"Resource","resource":{"name":"cpu","currentAverageUtilization":20,"currentAverageValue":"2m"}}]'
  creationTimestamp: 2018-01-15T04:49:25Z
  name: my-app
  namespace: default
  resourceVersion: "12404670"
  selfLink: /apis/autoscaling/v1/namespaces/default/horizontalpodautoscalers/my-app
  uid: 76718ca4-f9af-11e7-97d1-062eee34ec70
spec:
  maxReplicas: 3
  minReplicas: 1
  scaleTargetRef:
    apiVersion: extensions/v1beta1 --------->  It should have this line
    kind: Deployment
    name: my-app
  targetCPUUtilizationPercentage: 50
status:
  currentCPUUtilizationPercentage: 20
  currentReplicas: 1
  desiredReplicas: 1
  lastScaleTime: 2018-01-15T04:49:34Z
```

## Steps to Reproduce ##

1. Create a Pipeline with "Deploy" Stage for Kubernetes provider
2. Enable autoscaling and input the required fields 
3. Run pipeline
4. SSH to the deployment machine
5. Run "kubectl get hpa"

**Comment by axot** [comment]
Same issue happens in 1.9.x-gke, after adding `apiVersion: extensions/v1beta1` hpa works again.

**Comment by axot** [comment]
@antoooks I will push a PR to fix the issue, by the way, In my case,
in `HORIZONTAL POD AUTOSCALING`, it says `No autoscaler found`.

![hpa](https://user-images.githubusercontent.com/611054/37252527-a8223b62-2565-11e8-9fcd-08896438896a.jpg)

Did this also happened in your case? This issue may need be fixed too.

PS:
if I use a command below,  1HORIZONTAL POD AUTOSCALING1 will works anyway.
```
curl -X POST -H "Content-Type: application/json" -d  '[ {  "createServerGroup": { "application": "kub", "stack": "test",  "targetSize": "3", "securityGroups": [], "loadBalancers":  [],  "containers": [ { "name": "librarynginx", "imageDescription": { "repository": "library/nginx" } } ], "capacity": { "min": 1, "max": 5 }, "scalingPolicy": { "cpuUtilization": { "target": 40 } }, "account":  "my-kubernetes-account" } } ]' localhost:7002/kubernetes/ops
```

**Comment by axot** [comment]
> in `HORIZONTAL POD AUTOSCALING`, it says No autoscaler found.

This is my fault, after delete deployment annotations in RS, it works again.

```
static boolean hasDeployment(ReplicaSet replicaSet) {
    return replicaSet?.metadata?.annotations?.any { k, v -> k.startsWith(DEPLOYMENT_ANNOTATION) }
}
```

**Comment by antoooks** [comment]
@axot yup, that wasn't occured in my case.

glad you come to help with this issue, It's like untouched for like 2 months now...

**Comment by axot** [comment]
I think it just need a api version, for reference see this commit. 

This feature need k8s >= 1.9.0

https://github.com/kubernetes/kubernetes/commit/d2b41120ea0779954560e139a1dbcfa1bb36e088

**Comment by Wigder** [comment]
We encountered this issue and updated our version of Spinnaker to latest. I can confirm that our Clouddriver image has been updated to a version from the 19th of March, but still when we deploy new image of our software into GKE through our Spinnaker pipeline, the line "apiVersion: extensions/v1beta1" still isn't present when I run "kubectl edit hpa", so our HPA doesn't work until we manually add it. I can also confirm that we are using Kubernetes v1.9.6.
Any ideas on why this is happening?
